File: CS/XAF.LinqToXpo/LinqCollectionSourceProvider.cs
using System;
using System.Linq;
using DevExpress.Xpo;
using System.Reflection;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Xpo;
using System.Collections.Generic;

namespace XAF.LinqToXpo {
    public static class LinqCollectionSourceHelper {
        public static void CreateCustomCollectionSource(object sender, CreateCustomCollectionSourceEventArgs e) {
            IModelListViewLinq listViewInfo = ((XafApplication)sender).FindModelView(e.ListViewID) as IModelListViewLinq;
            if (listViewInfo == null) {
                return;
            }
            if (string.IsNullOrEmpty(listViewInfo.XPQueryMethod)) {
                return;
            }
            IQueryable query = LinqCollectionSourceHelper.InvokeMethod(e.ObjectType, listViewInfo.XPQueryMethod, ((XPObjectSpace)e.ObjectSpace).Session);
            if (query == null) {
                return;
            }
            e.CollectionSource = new LinqCollectionSource(e.ObjectSpace, e.ObjectType, query);
        }
        public static string[] GetXPQueryMethods(Type type) {
            List<string> names = new List<string>();
            MethodInfo[] methods = type.GetMethods(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Static);
            foreach (MethodInfo mi in methods) {
                if (IsCompatibleMethod(mi)) {
                    names.Add(mi.Name);
                }
            }
            return names.ToArray();
        }
        public static bool IsCompatibleMethod(MethodInfo mi) {
            ParameterInfo[] pis = mi.GetParameters();
            return mi.ReturnType != null && typeof(IQueryable).IsAssignableFrom(mi.ReturnType)
                    && pis.Length == 1 && pis[0].ParameterType.IsAssignableFrom(typeof(Session));
        }
        public static IQueryable InvokeMethod(Type type, string name, Session session) {
            MethodInfo method = FindMethod(type, name);
            if (method == null) {
                return null;
            }
            return (IQueryable)method.Invoke(null, new object[] { session });
        }
        private static MethodInfo FindMethod(Type type, string name) {
            MethodInfo[] methods = type.GetMethods(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Static);
            foreach (MethodInfo mi in methods) {
                if (mi.Name == name && IsCompatibleMethod(mi)) {
                    return mi;
                }
            }
            return null;
        }
        public static string[] GetDisplayableProperties(Type type, string name) {
            MethodInfo method = FindMethod(type, name);
            if (method == null) {
                return null;
            }
            foreach (QueryProjectionsAttribute attribute in method.GetCustomAttributes(typeof(QueryProjectionsAttribute), false)) {
                return attribute.Projections.Split(';');
            }
            return null;
        }
    }
    [AttributeUsage(AttributeTargets.Method)]
    public sealed class QueryProjectionsAttribute : Attribute {
        public QueryProjectionsAttribute(string projections) {
            this.Projections = projections;
        }
        //Comma-separated names of object properties to be included into the query results.
        public string Projections {
            get;
            private set;
        }
    }
}

File: VB/XAF.LinqToXpo/LinqCollectionSourceProvider.vb
Imports System
Imports System.Linq
Imports DevExpress.Xpo
Imports System.Reflection
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Xpo
Imports System.Collections.Generic

Namespace XAF.LinqToXpo
    Public NotInheritable Class LinqCollectionSourceHelper

        Private Sub New()
        End Sub

        Public Shared Sub CreateCustomCollectionSource(ByVal sender As Object, ByVal e As CreateCustomCollectionSourceEventArgs)
            Dim listViewInfo As IModelListViewLinq = TryCast(DirectCast(sender, XafApplication).FindModelView(e.ListViewID), IModelListViewLinq)
            If listViewInfo Is Nothing Then
                Return
            End If
            If String.IsNullOrEmpty(listViewInfo.XPQueryMethod) Then
                Return
            End If
            Dim query As IQueryable = LinqCollectionSourceHelper.InvokeMethod(e.ObjectType, listViewInfo.XPQueryMethod, CType(e.ObjectSpace, XPObjectSpace).Session)
            If query Is Nothing Then
                Return
            End If
            e.CollectionSource = New LinqCollectionSource(e.ObjectSpace, e.ObjectType, query)
        End Sub
        Public Shared Function GetXPQueryMethods(ByVal type As Type) As String()
            Dim names As New List(Of String)()
            Dim methods() As MethodInfo = type.GetMethods(System.Reflection.BindingFlags.Public Or System.Reflection.BindingFlags.Static)
            For Each mi As MethodInfo In methods
                If IsCompatibleMethod(mi) Then
                    names.Add(mi.Name)
                End If
            Next mi
            Return names.ToArray()
        End Function
        Public Shared Function IsCompatibleMethod(ByVal mi As MethodInfo) As Boolean
            Dim pis() As ParameterInfo = mi.GetParameters()
            Return mi.ReturnType IsNot Nothing AndAlso GetType(IQueryable).IsAssignableFrom(mi.ReturnType) AndAlso pis.Length = 1 AndAlso pis(0).ParameterType.IsAssignableFrom(GetType(Session))
        End Function
        Public Shared Function InvokeMethod(ByVal type As Type, ByVal name As String, ByVal session As Session) As IQueryable
            Dim method As MethodInfo = FindMethod(type, name)
            If method Is Nothing Then
                Return Nothing
            End If
            Return DirectCast(method.Invoke(Nothing, New Object() { session }), IQueryable)
        End Function
        Private Shared Function FindMethod(ByVal type As Type, ByVal name As String) As MethodInfo
            Dim methods() As MethodInfo = type.GetMethods(System.Reflection.BindingFlags.Public Or System.Reflection.BindingFlags.Static)
            For Each mi As MethodInfo In methods
                If mi.Name = name AndAlso IsCompatibleMethod(mi) Then
                    Return mi
                End If
            Next mi
            Return Nothing
        End Function
        Public Shared Function GetDisplayableProperties(ByVal type As Type, ByVal name As String) As String()
            Dim method As MethodInfo = FindMethod(type, name)
            If method Is Nothing Then
                Return Nothing
            End If
            For Each attribute As QueryProjectionsAttribute In method.GetCustomAttributes(GetType(QueryProjectionsAttribute), False)
                Return attribute.Projections.Split(";"c)
            Next attribute
            Return Nothing
        End Function
    End Class
    <AttributeUsage(AttributeTargets.Method)> _
    Public NotInheritable Class QueryProjectionsAttribute
        Inherits Attribute

        Public Sub New(ByVal projections As String)
            Me.Projections = projections
        End Sub
        'Comma-separated names of object properties to be included into the query results.
        Private privateProjections As String
        Public Property Projections() As String
            Get
                Return privateProjections
            End Get
            Private Set(ByVal value As String)
                privateProjections = value
            End Set
        End Property
    End Class
End Namespace
